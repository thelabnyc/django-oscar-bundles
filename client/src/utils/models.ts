import * as t from 'io-ts';
import {failure} from 'io-ts/lib/PathReporter';
import {Either, isLeft} from 'fp-ts/lib/Either';


export const check = <T>(result: Either<t.Errors, T>): T => {
    if (isLeft(result)) {
        throw new Error(failure(result.left).join('\n'));
    }
    return result.right;
};

const nullable = <RT extends t.Mixed>(type: RT) => {
    return t.union([t.null, type]);
};

export const optional = <RT extends t.Mixed>(type: RT) => {
    return t.union([t.undefined, type]);
};

class FileType extends t.Type<File> {
    readonly _tag: 'NumberType' = 'NumberType';
    constructor() {
        super(
            'number',
            (f): f is File => {
                return f instanceof File;
            },
            (u, c) => {
                return this.is(u) ? t.success(u) : t.failure(u, c);
            },
            t.identity
        );
    }
}
const file = new FileType();



export const DRFSelectOption = t.interface({
    value: t.string,
    display_name: t.string,
});
export const DRFSelectOptions = t.readonlyArray(DRFSelectOption);

export const BaseField = <T extends string>(fieldType: t.LiteralC<T>) => {
    return t.interface({
        type: fieldType,
        required: t.boolean,
        read_only: t.boolean,
        label: t.string,
        help_text: optional(t.string),
    });
};

export const IntegerField = BaseField(t.literal("integer"));

export const BooleanField = BaseField(t.literal("boolean"));

export const DatetimeField = BaseField(t.literal("datetime"));

export const StringField = t.intersection([
    BaseField(t.literal("string")),
    t.interface({
        max_length: optional(t.number),
    }),
]);

export const ChoiceField = t.intersection([
    BaseField(t.literal("choice")),
    t.interface({
        choices: DRFSelectOptions,
    }),
]);

export const Field = t.union([
    IntegerField,
    BooleanField,
    DatetimeField,
    StringField,
    ChoiceField,
    t.unknown,
]);

export const HTTPMethod = t.union([
    t.literal("GET"),
    t.literal("HEAD"),
    t.literal("OPTIONS"),
    t.literal("POST"),
    t.literal("PUT"),
    t.literal("PATCH"),
    t.literal("DELETE"),
]);

export const DRFOptionsResponse = t.interface({
    name: t.string,
    description: t.string,
    renders: t.readonlyArray(t.string),
    parses: t.readonlyArray(t.string),
    actions: t.interface({
        GET: optional(t.record(t.string, Field)),
        HEAD: optional(t.record(t.string, Field)),
        OPTIONS: optional(t.record(t.string, Field)),
        POST: optional(t.record(t.string, Field)),
        PUT: optional(t.record(t.string, Field)),
        PATCH: optional(t.record(t.string, Field)),
        DELETE: optional(t.record(t.string, Field)),
    }),
});




export const Product = t.interface({
    id: t.number,
    dashboard_url: t.string,
    product_class: nullable(t.interface({
        id: t.number,
        name: t.string,
    })),
    title: t.string,
    slug: t.string,
    is_parent: t.boolean,
    is_child: t.boolean,
    parent: nullable(t.number),
    children: t.array(t.number),
});
export const Products = t.array(Product);



export const Range = t.interface({
    id: t.number,
    dashboard_url: t.string,
    name: t.string,
    slug: t.string,
    description: t.string,
});
export const Ranges = t.array(Range);



export const ConcreteBundle = t.interface({
    id: nullable(t.number),
    triggering_product: t.number,
    suggested_products: t.array(t.number),
});
export const ConcreteBundles = t.array(ConcreteBundle);



export const UserConfigurableBundle = t.interface({
    id: nullable(t.number),
    triggering_product: t.number,
    suggested_range: t.number,
    quantity: t.number,
});
export const UserConfigurableBundles = t.array(UserConfigurableBundle);



export const BundleGroup = t.interface({
    id: nullable(t.number),
    bundle_type: t.string,
    name: t.string,
    description: t.string,
    headline: t.string,
    image: optional(nullable(t.string)),
    newImage: optional(nullable(file)),
    clearImage: optional(t.boolean),
    triggering_parents: t.array(t.number),
    suggested_parents: t.array(t.number),
    concrete_bundles: t.array(ConcreteBundle),
    user_configurable_bundles: t.array(UserConfigurableBundle),
});
export const BundleGroups = t.array(BundleGroup);
