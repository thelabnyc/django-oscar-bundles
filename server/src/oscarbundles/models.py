from django.db import models
from django.conf import settings
from django.utils.translation import ugettext_lazy as _
from oscar.core.loading import get_model

Product = get_model('catalogue', 'Product')


class BundleGroup(models.Model):
    GROUP_TYPE_OPTIONS = getattr(settings, 'BUNDLE_GROUP_TYPES', (
        ('default', 'Default'),
    ))

    bundle_type = models.CharField(_('Bundle Type'),
        max_length=200,
        default=GROUP_TYPE_OPTIONS[0][0],
        choices=GROUP_TYPE_OPTIONS)

    name = models.CharField(_('Name'),
        max_length=200,
        blank=True,
        default='')
    headline = models.CharField(
        max_length=200,
        blank=True,
        default='Forget Something?',
        help_text='CTA headline in cart display')
    description = models.TextField(_('Description'),
        blank=True,
        default='')
    image = models.ImageField(
        upload_to=settings.OSCAR_IMAGE_FOLDER,
        max_length=255,
        blank=True,
        null=True)

    triggering_parents = models.ManyToManyField('catalogue.Product',
        related_name='triggering_bundle_groups',
        verbose_name=_('Triggering Products'))
    suggested_parents = models.ManyToManyField('catalogue.Product',
        related_name='suggesting_bundle_groups',
        verbose_name=_('Suggested Products'))


class Bundle(models.Model):
    bundle_group = models.ForeignKey(BundleGroup,
        on_delete=models.CASCADE,
        related_name='bundles')

    triggering_product = models.ForeignKey('catalogue.Product',
        on_delete=models.CASCADE,
        related_name='triggering_bundles',
        verbose_name=_('Triggering Product'),
        help_text=_('Which product should trigger this bundle?'))
    suggested_products = models.ManyToManyField('catalogue.Product',
        related_name='suggesting_bundles',
        verbose_name=_('Suggested Products'),
        help_text=_('Which product(s) should this bundle suggest when triggered?'))

    class Meta:
        unique_together = (
            ('bundle_group', 'triggering_product'),
        )
