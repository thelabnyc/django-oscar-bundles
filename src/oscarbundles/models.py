from django.db import models
from django.conf import settings
from django.utils.translation import ugettext_lazy as _
from oscar.core.loading import get_model

Product = get_model('catalogue', 'Product')


class BundleMetaMixin(models.Model):
    name = models.CharField(_('Name'),
        max_length=200,
        blank=True,
        default='',
        help_text=_('Optional name for this bundle'))
    description = models.TextField(_('Description'),
        blank=True,
        default='',
        help_text=_('Optional description for this bundle'))
    image = models.ImageField(
        upload_to=settings.OSCAR_IMAGE_FOLDER, max_length=255,
        blank=True, null=True)

    class Meta:
        abstract = True

    def __str__(self):
        return self.name


class Bundle(BundleMetaMixin, models.Model):
    bundle_group = models.ForeignKey('BundleGroup',
        blank=True, null=True, help_text=_('Optional grouping for bundle'), on_delete=models.CASCADE)
    triggering_products = models.ManyToManyField('catalogue.Product',
        related_name='triggering_bundles',
        verbose_name=_('Triggering Products'),
        help_text=_('Which product(s) should trigger this bundle?'))
    suggested_products = models.ManyToManyField('catalogue.Product',
        related_name='suggesting_bundles',
        verbose_name=_('Suggested Products'),
        help_text=_('Which product(s) should this bundle suggest when triggered?'))


class BundleGroup(BundleMetaMixin, models.Model):
    pass
