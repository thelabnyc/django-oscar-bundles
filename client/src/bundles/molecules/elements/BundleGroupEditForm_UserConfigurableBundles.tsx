import React from "react";
import Select from "react-select";
import {
    IProduct,
    IRange,
    SelectOption,
} from "../../../utils/models.interfaces";

type ILinkedRange = {
    rangeID: number | null;
    quantity: number;
};
export type ILinkedRanges = {
    [triggeringProductID: number]: ILinkedRange[];
};

interface IProps {
    ranges: IRange[];
    isSaving: boolean;
    triggeringParents: number[];
    linkedRanges: ILinkedRanges;
    getProduct: (productID: number) => IProduct;
    onLinkedRangesChange: (
        trigger: IProduct,
        rangeIndex: number,
        opts: SelectOption | ReadonlyArray<SelectOption> | null | undefined
    ) => void;
    onLinkedRangeQuantityChange: (
        trigger: IProduct,
        rangeIndex: number,
        opts: SelectOption | ReadonlyArray<SelectOption> | null | undefined
    ) => void;
}

interface IState {}

const helpText =
    gettext(`User configurable bundles are bundles which link parent/standalone products to ranges of \
other products. Their intent is to represent a group of products related to the suggesting parent, rather than a \
specific one-to-one mapping of product to product.`);

export class UserConfigurableBundles extends React.PureComponent<
    IProps,
    IState
> {
    private getRangeSelectOptions() {
        const rangeOptions = this.props.ranges
            .sort((a, b) => {
                return a.name.localeCompare(b.name);
            })
            .map((p) => {
                const opt: SelectOption = {
                    value: p.id,
                    label: p.name,
                };
                return opt;
            });
        return rangeOptions;
    }

    private getQuantitySelectOptions() {
        const quantityOptions: SelectOption[] = [];
        for (let i = 1; i < 100; i++) {
            quantityOptions.push({
                value: i,
                label: `${i}`,
            });
        }
        return quantityOptions;
    }

    private buildBundleRows(triggerParent: IProduct) {
        const suggestOptions = this.getRangeSelectOptions();
        const quantityOptions = this.getQuantitySelectOptions();
        const linkedRanges = this.props.linkedRanges[triggerParent.id] || [];
        if (
            linkedRanges.length <= 0 ||
            linkedRanges[linkedRanges.length - 1].rangeID !== null
        ) {
            linkedRanges.push({
                rangeID: null,
                quantity: 1,
            });
        }
        return (
            <table
                key={triggerParent.id}
                className="table bundle-group-edit__link-table"
            >
                <tbody>
                    {linkedRanges.map((range, i) => {
                        return (
                            <tr key={i}>
                                <th>
                                    <a
                                        href={triggerParent.dashboard_url}
                                        target="_blank"
                                    >
                                        {triggerParent.title}
                                    </a>
                                </th>
                                <td>
                                    <Select
                                        isClearable={true}
                                        value={suggestOptions.filter(
                                            (o) => o.value === range.rangeID
                                        )}
                                        options={suggestOptions}
                                        onChange={this.props.onLinkedRangesChange.bind(
                                            this,
                                            triggerParent,
                                            i
                                        )}
                                        isDisabled={this.props.isSaving}
                                    />
                                </td>
                                <td>
                                    <Select
                                        value={quantityOptions.filter(
                                            (o) =>
                                                range.rangeID !== null &&
                                                o.value === range.quantity
                                        )}
                                        options={quantityOptions}
                                        onChange={this.props.onLinkedRangeQuantityChange.bind(
                                            this,
                                            triggerParent,
                                            i
                                        )}
                                        isDisabled={
                                            this.props.isSaving ||
                                            range.rangeID === null
                                        }
                                    />
                                </td>
                            </tr>
                        );
                    })}
                </tbody>
            </table>
        );
    }

    private buildParentRows() {
        if (this.props.triggeringParents.length <= 0) {
            return (
                <div className="row">
                    <div className="col-sm-12">
                        <p>
                            <em>
                                {gettext(
                                    "Select at least one trigger product to begin linking ranges."
                                )}
                            </em>
                        </p>
                    </div>
                </div>
            );
        }
        return (
            <div className="row">
                <div className="col-sm-4">
                    <h3>
                        <strong>{gettext("Trigger Parent")}</strong>
                    </h3>
                </div>
                <div className="col-sm-4">
                    <h3>
                        <strong>{gettext("Suggested Range")}</strong>
                    </h3>
                </div>
                <div className="col-sm-4">
                    <h3>
                        <strong>{gettext("Suggested Quantity")}</strong>
                    </h3>
                </div>
                <div className="col-sm-12">
                    {this.props.triggeringParents.map((triggerParentID) => {
                        const triggerParent =
                            this.props.getProduct(triggerParentID);
                        return this.buildBundleRows(triggerParent);
                    })}
                </div>
            </div>
        );
    }

    render() {
        return (
            <div className="col-sm-12 bundle-group-edit__section">
                <h1>{gettext("Range Links (User Configurable Bundles)")}</h1>
                <p>
                    <em>{helpText}</em>
                </p>
                {this.buildParentRows()}
            </div>
        );
    }
}
