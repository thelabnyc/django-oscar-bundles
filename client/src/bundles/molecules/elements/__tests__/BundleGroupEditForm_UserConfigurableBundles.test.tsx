import React from "react";
import { render } from "@testing-library/react";
import { IProduct, IRange } from "../../../../utils/models.interfaces";
import { UserConfigurableBundles } from "../BundleGroupEditForm_UserConfigurableBundles";

const products: IProduct[] = [
    {
        id: 1,
        dashboard_url: "http://localhost/store/dashboard/catalogue/products/1/",
        product_class: {
            id: 1,
            name: "books",
        },
        title: "book-1",
        slug: "book-1",
        is_parent: false,
        is_child: false,
        parent: null,
        children: [],
    },
    {
        id: 2,
        dashboard_url: "http://localhost/store/dashboard/catalogue/products/2/",
        product_class: {
            id: 1,
            name: "books",
        },
        title: "book-2",
        slug: "book-2",
        is_parent: false,
        is_child: false,
        parent: null,
        children: [],
    },
];

const ranges: IRange[] = [
    {
        id: 1,
        dashboard_url: "http://localhost/store/dashboard/ranges/1/",
        name: "All Products",
        slug: "all-products",
        description: "",
    },
    {
        id: 2,
        dashboard_url: "http://localhost/store/dashboard/ranges/2/",
        name: "Books",
        slug: "books",
        description: "",
    },
];

describe("#bundles/molecules/elements/UserConfigurableBundles", () => {
    it("renders", () => {
        const wrapper = render(
            <UserConfigurableBundles
                ranges={ranges}
                isSaving={false}
                triggeringParents={[1]}
                linkedRanges={{
                    "1": [
                        {
                            rangeID: 1,
                            quantity: 2,
                        },
                    ],
                }}
                getProduct={() => products[0]}
                onLinkedRangesChange={() => null}
                onLinkedRangeQuantityChange={() => null}
            />
        );
        expect(wrapper.asFragment()).toMatchSnapshot();
    });
});
