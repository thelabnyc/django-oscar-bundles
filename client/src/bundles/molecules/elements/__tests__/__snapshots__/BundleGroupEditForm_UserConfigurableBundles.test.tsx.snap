// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#bundles/molecules/elements/UserConfigurableBundles renders 1`] = `
<DocumentFragment>
  <div
    class="col-sm-12 bundle-group-edit__section"
  >
    <h1>
      Range Links (User Configurable Bundles)
    </h1>
    <p>
      <em>
        User configurable bundles are bundles which link parent/standalone products to ranges of other products. Their intent is to represent a group of products related to the suggesting parent, rather than a specific one-to-one mapping of product to product.
      </em>
    </p>
    <div
      class="row"
    >
      <div
        class="col-sm-4"
      >
        <h3>
          <strong>
            Trigger Parent
          </strong>
        </h3>
      </div>
      <div
        class="col-sm-4"
      >
        <h3>
          <strong>
            Suggested Range
          </strong>
        </h3>
      </div>
      <div
        class="col-sm-4"
      >
        <h3>
          <strong>
            Suggested Quantity
          </strong>
        </h3>
      </div>
      <div
        class="col-sm-12"
      >
        <table
          class="table bundle-group-edit__link-table"
        >
          <tbody>
            <tr>
              <th>
                <a
                  href="http://localhost/store/dashboard/catalogue/products/1/"
                  target="_blank"
                >
                  book-1
                </a>
              </th>
              <td>
                <div
                  class=" css-2b097c-container"
                >
                  <div
                    class=" css-yk16xz-control"
                  >
                    <div
                      class=" css-g1d714-ValueContainer"
                    >
                      <div
                        class=" css-1uccc91-singleValue"
                      >
                        All Products
                      </div>
                      <div
                        class="css-b8ldur-Input"
                      >
                        <div
                          class=""
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            id="react-select-2-input"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                      <span
                        class=" css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </td>
              <td>
                <div
                  class=" css-2b097c-container"
                >
                  <div
                    class=" css-yk16xz-control"
                  >
                    <div
                      class=" css-g1d714-ValueContainer"
                    >
                      <div
                        class=" css-1uccc91-singleValue"
                      >
                        2
                      </div>
                      <div
                        class="css-b8ldur-Input"
                      >
                        <div
                          class=""
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            id="react-select-3-input"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class=" css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <th>
                <a
                  href="http://localhost/store/dashboard/catalogue/products/1/"
                  target="_blank"
                >
                  book-1
                </a>
              </th>
              <td>
                <div
                  class=" css-2b097c-container"
                >
                  <div
                    class=" css-yk16xz-control"
                  >
                    <div
                      class=" css-g1d714-ValueContainer"
                    >
                      <div
                        class=" css-1wa3eu0-placeholder"
                      >
                        Select...
                      </div>
                      <div
                        class="css-b8ldur-Input"
                      >
                        <div
                          class=""
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            id="react-select-4-input"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class=" css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </td>
              <td>
                <div
                  class=" css-14jk2my-container"
                >
                  <div
                    class=" css-1fhf3k1-control"
                  >
                    <div
                      class=" css-g1d714-ValueContainer"
                    >
                      <div
                        class=" css-1wa3eu0-placeholder"
                      >
                        Select...
                      </div>
                      <div
                        class="css-b8ldur-Input"
                      >
                        <div
                          class=""
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            disabled=""
                            id="react-select-5-input"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class=" css-109onse-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
